feat: 로그인/회원가입 모달 ui 구현

mj0107 opened the pull request:
## 📋 변경 사항

- 로그인/회원가입 모달 폼 ui 구현
- react-hook-form과 zod 이용해서 폼 유효성 검사
- 소셜 로그인 위해서 각 플랫폼 별 아이콘 이미지 삽입

## 🏷️ PR 유형

- [x] 🚀 feat: 새로운 기능 추가
- [ ] 🐛 fix: 버그 수정
- [x] 🎨 design: UI/UX 디자인 변경
- [ ] ♻️ refactor: 코드 리팩토링
- [ ] 📝 docs: 문서 수정
- [ ] ✅ test: 테스트 코드 추가/수정
- [x] 🔧 chore: 빌드/패키지 매니저 수정
- [ ] 🎨 style: 코드 포맷팅 (기능 변경 없음)

## ✅ 체크리스트

- [x] 커밋 메시지 컨벤션 준수
- [x] 로컬 테스트 완료
- [x] 코드 리뷰 준비 완료

## 📸 테스트 결과

<img width="1511" height="822" alt="스크린샷 2025-07-18 오후 1 24 16" src="https://github.com/user-attachments/assets/8d2a6637-10f8-44b5-b752-9d95c8d2de0c" />
<img width="1511" height="822" alt="스크린샷 2025-07-18 오후 1 24 33" src="https://github.com/user-attachments/assets/81f7654d-30f8-4aca-b0a0-2e92bd02b1ec" />
<img width="1511" height="822" alt="스크린샷 2025-07-18 오후 1 25 04" src="https://github.com/user-attachments/assets/005ac19c-5cf4-4151-838b-db3acc40b6c4" />


## 🔗 관련 이슈

- Closes #12

mj0107 committed:
chore: react-hook-form 의존성 설치 #12

mj0107 committed:
design: 소셜 로그인 로고 이미지 추가 #12

- GitHub, Google, Naver

mj0107 committed:
feat: 소셜 로그인 로고 컴포넌트 구현 #12

mj0107 committed:
refactor: LoginModal에서 AuthModal로 변수명 변경 #12

mj0107 committed:
feat: 로그인/회원가입 form 구현 #12

- react-hook-form 사용하여 구현

mj0107 committed:
feat: 로그인/회원가입 모달 구현 #12

- shadcn/ui Dialog 이용해서 구현
- 로그인/회원가입 상태 전환 구현

genre0928 commented on src/components/auth-modal/auth-form/email/ErrorMessage.tsx:
```diff
@@ -0,0 +1,27 @@
+  errors: FieldErrors<{ email: string }>;
+}
+
+export default function ErrorMessage({ control, errors }: ErrorMessageProps) {
+  return (
+    <>
+      {/* 필드 에러 메시지 */}
+      <FormField
+        control={control}
+        name="email"
+        render={() => <FormMessage />}
+      />
+
+      {/* 서버 에러 메시지 */}
+      {errors.root?.server && (
+        <p className="text-sm text-red-500" role="alert">
+          {errors.root.server.message}
+        </p>
+      )}
+    </>
+  );
+}
```
혹시 여기 ErrorMessage 컴포넌트를 분리한 이유가 있을까요?!
직전의 EmailInput 컴포넌트에서 react-hook-form을 통해 데이터 검증과 에러메시지 렌더링을 하는 것 같은데 동일한 기능이 추가된 것 같아서 여쭤봐요

그리고 만약 입력란이 여러개가 되는 경우라면 공통 컴포넌트화해서 재사용을 하는 방식으로 분리하는거겠죠??

mj0107 replied:
일단 목적은 재사용성을 위한 분리가 목적이 맞습니다.

다만, Emailinput 컴포넌트에선 에러메시지 렌더링을 하지 않고 있습니다.
```typescript
<FormField
      control={control}
      name="email"
      render={({ field }) => (
        <FormItem className="flex-1">
          <FormControl>
            <Input
              type="email"
              placeholder="이메일을 입력하세요."
              {...field}
              className="shadow-none rounded-none placeholder:text-tertiary placeholder:text-md h-12 focus-visible:ring-0 focus:border-green focus-visible:border-green"
              aria-invalid={!!errors.email}
            />
          </FormControl>
        </FormItem>
      )}
    />
```
아마 `aria-invalid={!!errors.email}`부분을 착각하신 것 같은데,
`aria-invalid`는 시각장애인을 위한 속성입니다.
aria에는 invalid 말고 다양한 옵션들이 있는데,
여기서 invalid가 true가 되면, 스크린 리더가 "이 필드는 유효하지 않음" 과 같은 음성으로 알려주게 됩니다.
웹 접근성을 위한 속성이라고 보시면 될 것 같아요.

mj0107 replied:
근데 적다보니, ErrorMessage에는 서버 에러 메시지 부분만 남겨두고,
해당 input에 관한 에러는 그냥 컴포넌트에 같이 묶는게 좋겠네요...
요건 EmailInput에 FormMessage 컴포넌트 추가해서 수정하겠습니다.

genre0928 replied:
웹사이트를 사용하다 보면 input 요소에 대해서 입력한 데이터가 검증되지 않은 경우 input 하단에 에러 메시지를 보여주는 방식이 좋았던 기억이 있어요!
그래서 저는 UX의 관점에서 react-hook-form 라이브러리를 사용했었거든요!!
그리구 에러핸들링을 한다고 착각했던 부분이 {...field} 였어요!
Input 태그에 전달되는 Props인데 렌더링하는 부분으로 잘못 봤네요 ㅎㅎㅎ

컴포넌트 수정에 대한 부분도 확인 완료입니다!!

genre0928 commented on src/components/auth-modal/auth-form/email/index.tsx:
```diff
@@ -0,0 +1,72 @@
+    },
+  });
+
+  // 폼 제출 성공 핸들러
+  const onSubmit: SubmitHandler<AuthFormData> = async (data) => {
+    console.log(data);
+  };
+
+  // 폼 제출 에러 핸들러
+  const onError: SubmitErrorHandler<AuthFormData> = (errors) => {
+    console.error("폼 검증 실패:", errors);
+  };
```
여기는 API 연동 확인 때문에 디버깅 코드 그대로 냅두신거죠?!

mj0107 replied:
> 여기는 API 연동 확인 때문에 디버깅 코드 그대로 냅두신거죠?!

넵 이거 바로 다음에 api 연결할 것 같아서요!

genre0928 replied:
넵 확인 완료입니다!
이전 제 코드 리뷰에서 디버깅 코드에 대한 부분 리뷰해주신 것 때문에 확인 차 여쭤봤습니다!!

genre0928 commented on src/components/auth-modal/index.tsx:
```diff
@@ -0,0 +1,35 @@
+}
+
+export default function AuthModal({ open, onOpenChange }: AuthModalProps) {
+  const [mode, setMode] = useState<"login" | "signup">("login");
+
+  const isLoginMode = mode === "login";
+
+  const handleModeChange = (mode: "login" | "signup") => {
+    setMode(mode);
+  };
```
mode 변수의 타입은 자주 등장하는 것 같은데 types에 따로 분리해서 관리하는 방법이 아니라 직접 작성하신 것은 의도한 바일까요???

mj0107 replied:
> mode 변수의 타입은 자주 등장하는 것 같은데 types에 따로 분리해서 관리하는 방법이 아니라 직접 작성하신 것은 의도한 바일까요???

따로 해도 되긴하는데, 어차피 로그인/회원가입 두 모드밖에 없어서 그냥 넣어놨습니다!

genre0928 replied:
넵넵! 확인했습니다

genre0928 commented on src/components/ui/dialog.tsx:
```diff
@@ -38,7 +38,7 @@ function DialogOverlay({
       data-slot="dialog-overlay"
       className={cn(
-        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 fixed inset-0 z-50 bg-black/50",
+        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 fixed inset-0 z-50 bg-content/80",
```
저는 shadcn의 컴포넌트를 사용할 때 원래 설정된 속성을 무시하거나 변경할 때 !p-x 이런 식으로 작성을 많이 하는데요
왜냐하면 import한 컴포넌트가 다른 데서도 쓰일 때 그대로 활용되길 바랄 수도 있고, 의도하지 않은 CSS가 반영될 가능성을 고려해서요!

민준님은 공통 요소 변경하시는 방식을 선호하시는거죠?

mj0107 replied:
> 저는 shadcn의 컴포넌트를 사용할 때 원래 설정된 속성을 무시하거나 변경할 때 !p-x 이런 식으로 작성을 많이 하는데요 왜냐하면 import한 컴포넌트가 다른 데서도 쓰일 때 그대로 활용되길 바랄 수도 있고, 의도하지 않은 CSS가 반영될 가능성을 고려해서요!
> 
> 민준님은 공통 요소 변경하시는 방식을 선호하시는거죠?

이 부분이 로그인 모달 뜰 때 뒤에 깔리는 배경, 즉 backdrop 부분(shadcn에선 overlay)인데요,
이 모달 backdrop은 다른 모달이 있다고 해도, 각 모달마다 다르게 지정할 이유는 없을거라고 판단했습니다.
그래서 기본 스타일을 다르게 변경해줬어요!

만약, shadcn 컴포넌트가 각 사용처마다 스타일이 다르다면,
해당 사용처에서 말씀해주신것처럼 각각 스타일을 지정하는 방식이 맞습니다.
지금은 어차피 통일된 스타일을 가질거기 때문에,
기본 스타일을 바꿔줬습니다!

genre0928 replied:
넵! 공통 컴포넌트의 속성을 수정한 의도가 있는거였네용
확인 완료입니다!!!

genre0928 commented on src/components/ui/form.tsx:
```diff
@@ -0,0 +1,168 @@
+
+const Form = FormProvider;
+
+type FormFieldContextValue<
+  TFieldValues extends FieldValues = FieldValues,
+  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
+> = {
+  name: TName;
+};
```
혹시 여기 왜 제너릭 타입으로 타입 선언이 된 지 알 수 있을까요?
전달받는 값이 입력될 때 결정돼야 제너릭 타입을 쓴다고 알고 있는데 내부에 작성한 게 어떤 의도인 질 잘 모르겠어가지구요!
코드가 너무 어렵읍니다,,,,!!

mj0107 replied:
> 혹시 여기 왜 제너릭 타입으로 타입 선언이 된 지 알 수 있을까요? 전달받는 값이 입력될 때 결정돼야 제너릭 타입을 쓴다고 알고 있는데 내부에 작성한 게 어떤 의도인 질 잘 모르겠어가지구요! 코드가 너무 어렵읍니다,,,,!!

일단 한 마디로 정리하자면,
**재사용성을 위한 라이브러리 코드이기 때문에,
여러 폼 형태에 따라서 타입 추론이 유연하게 열려있다.**
라고 생각하시면 될 것 같습니다.

엄,,, 일단 이 코드는 shadcn에서 react-hook-form을 설치할 때 자동으로 삽입되는 코드입니다.
저도 자세히 들여다보진 않아서 정확하진 않지만

그래도 설명드리자면, 
`TFieldValues`는 전체 폼 데이터의 타입입니다.
폼에 어떠한 요소들이 있을지 알 수 없기 때문에 자동 추론을 위해 제네릭으로 설정한거죠.
물론, 지금은 비밀번호 input 하나만 있기 때문에 지정해주면 되겠죠?
하지만 이 코드는 라이브러리 코드이고, 재사용성을 위해 확장성에 초점을 맞췄겠죠.
어떠한 폼 형태가 들어오더라도 사용할 때 유연하게 타입을 받을 수 있도록 한거예요.

`TName`은 특정 필드 이름의 타입이에요.
```typescript
    <FormField
      control={control}
      name="email"
      render={({ field }) => (
        <FormItem className="flex-1">
          <FormControl>
            <Input
              type="email"
              placeholder="이메일을 입력하세요."
              {...field}
              className="shadow-none rounded-none placeholder:text-tertiary placeholder:text-md h-12 focus-visible:ring-0 focus:border-green focus-visible:border-green"
              aria-invalid={!!errors.email}
            />
          </FormControl>
        </FormItem>
      )}
    />
``` 
EamilInput 컴포넌트에서 위와 같이 사용했는데,
name 속성에 설정되어있는 'email'이 TName입니다.
위에 있는 control에서 타입이 자동으로 추론되는데,
name에 'email'을 설정하고, 추론한것과 맞는지 자동으로 검증되도록 한거예요.

Input 태그에 type을 email로 설정했죠?
이게 control에서 자동으로 추론되고,
name에서 검증하는겁니다.
'emaill'처럼 오타가 나면 틀리는거구요.

genre0928 replied:
아하 제너릭 타입을 의도해서 작성하신 줄 알고 사용하신 의도가 있나 했었는데,, 전혀 잘못 생각했었네요 ㅎㅎㅎ
그래도 라이브러리에서 재사용성을 위해서 제너릭 타입으로 선언한 이유는 확인했습니다,,!

추가로 control Props가 어떤 역할을 하는 지도 알게되었읍니다,,,,!!

mj0107 committed:
refactor: 에러메시지 역할 분리 #12

- 기존 ErrorMessage 컴포넌트는 ServerErrorMessage로 변경하여, 서버 에러만 보여주도록 수정
- 각 필드의 에러는 각 필드 컴포넌트에서 같이 보여주도록 수정하여 응집도 높임

mj0107 committed:
Merge branch 'develop' into feature/auth-modal-ui
